src/adapter/src/coord.rs

```diff
@@ -1487,6 +1495,77 @@ impl Coordinator {
+    ) -> Antichain<Timestamp> {
+        // All inputs must be readable at the chosen `as_of`, so it must be at least the join of
+        // the `since`s of all dependencies.
+        let storage_ids = dataflow.source_imports.keys().copied().collect();
+        let compute_ids = dataflow.index_imports.keys().copied().collect();
+        let id_bundle = CollectionIdBundle {
+            storage_ids,
+            compute_ids: btreemap! {cluster_id => compute_ids},
+        };
+        let mut as_of = self.least_valid_read(&id_bundle);
+
+        // For compute reconciliation to recognize that an existing dataflow can be reused, we want
+        // to advance the `as_of` far enough that it is beyond the `as_of`s of all dataflows that
+        // might still be installed on replicas, but ideally not much farther as that would just
+        // increase the wait time until the index becomes readable. We advance the `as_of` to the
+        // meet of the `upper`s of all dependencies, as we know that no replica can have produced
+        // output for that time, so we can assume that no replica `as_of` has been adanced beyond
+        // this time either.
+        let write_frontier = self.least_valid_write(&id_bundle);
+        // Things go wrong if we try to create a dataflow with `as_of = []`, so avoid that.
+        if !write_frontier.is_empty() {
+            as_of.join_assign(&write_frontier);
+        }
+
+        as_of
+    }
+
+    /// Returns an `as_of` suitable for bootstrapping the given materialized view dataflow.
+    fn bootstrap_materialized_view_as_of(
+        &self,
+        cluster_id: ComputeInstanceId,
+        dataflow: &DataflowDescription<OptimizedMirRelationExpr>,
+    ) -> Antichain<Timestamp> {
+        // All inputs must be readable at the chosen `as_of`, so it must be at least the join of
+        // the `since`s of all dependencies.
+        let storage_ids = dataflow.source_imports.keys().copied().collect();
+        let compute_ids = dataflow.index_imports.keys().copied().collect();
+        let id_bundle = CollectionIdBundle {
+            storage_ids,
+            compute_ids: btreemap! {cluster_id => compute_ids},
+        };
+        let mut as_of = self.least_valid_read(&id_bundle);
+
+        // For compute reconciliation to recognize that an existing dataflow can be reused, we want
+        // to advance the `as_of` as far as possible. If a storage collection for the MV already
+        // exists, we can advance to that collection's upper. This is the most we can advance the
+        // `as_of` without skipping times in the MV output.
+        let sink_id = dataflow
+            .sink_exports
+            .keys()
+            .exactly_one()
+            .expect("MV dataflow must export a sink");
+        let write_frontier = &self
+            .controller
+            .storage
+            .collection(*sink_id)
+            .expect("storage collection must exist")
+            .write_frontier;
```

Comment:
This seems to be very similar to `least_valid_write` with an `id_bundle` of `sink_id`. Might use it to avoid code duplication?

Reply:
That would turn out like this:
```rust
        let sink_id = dataflow
            .sink_exports
            .keys()
            .exactly_one()
            .expect("MV dataflow must export a sink");
        let id_bundle = CollectionIdBundle {
            storage_ids: [*sink_id].into(),
            compute_ids: Default::default(),
        };
        let write_frontier = self.least_valid_write(&id_bundle);
```

I think this is less obvious than getting the write frontier from the storage controller directly.

However, looking at `least_valid_write` I realized that I can get the storage frontier using `self.storage_write_frontier(*sink_id)`, which is indeed nicer!